src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/Dependency.cs

```diff
@@ -102,13 +103,9 @@ public Dependency(IDependencyModel dependencyModel, ITargetFramework targetFrame
             }
             else
             {
-                ImmutableList<string>.Builder normalizedDependencyIDs = ImmutableList.CreateBuilder<string>();
-                foreach (string id in dependencyModel.DependencyIDs)
-                {
-                    normalizedDependencyIDs.Add(GetID(TargetFramework, ProviderType, id));
-                }
-
-                DependencyIDs = normalizedDependencyIDs.ToImmutable();
+                IEnumerable<string> ids = dependencyModel.DependencyIDs.Select(
+                    id => GetID(TargetFramework, ProviderType, id));
+                DependencyIDs = ImmutableArray.CreateRange(ids);
```

Comment:
Could we use an `ImmutableArrayBuilder` here, and pre-size it with `dependencyModel.DependencyIDs.Count`?

Reply:
I couldn't find an approach here that didn't involve redundant copying. The builder always does a defensive copy in the end. The closest was an `ImmutableArray.CreateRange` overload that takes a func for mapping (so can create the backing array internally without defensive copies), but it only accepts `ImmutableArray` as an argument, and `DependencyIds` is `IImmutableList<string>`. It would be possible to [add such a method](https://github.com/dotnet/corefx/issues/32839) to `ImmutableArray`.

Even with the redundant copies, the builder is still faster and allocates less than `AddRange` 👍 

|                             Method | Count |       Mean |     Error |     StdDev |  Gen 0 | Allocated |
|----------------------------------- |------ |-----------:|----------:|-----------:|-------:|----------:|
|         ImmutableArray_CreateRange |     3 |   470.1 ns | 10.096 ns |  16.013 ns | 0.0196 |     104 B |
|     ImmutableArray_Builder |     3 |   107.2 ns |  1.483 ns |   1.315 ns | 0.0122 |      64 B |
|         ImmutableArray_CreateRange |    30 | 1,634.1 ns | 32.726 ns |  83.297 ns | 0.0896 |     472 B |
|     ImmutableArray_Builder |    30 |   585.9 ns | 11.733 ns |  18.609 ns | 0.0525 |     280 B |

Reply:
MoveToImmutable avoids a copy.

Reply:
Very nice. Updated.

|                                 Method | Count |        Mean |      Error |      StdDev |  Gen 0 | Allocated |
|--------------------------------------- |------ |------------:|-----------:|------------:|-------:|----------:|
|             ImmutableArray_CreateRange |     3 |   473.69 ns |  9.3715 ns |  10.0274 ns | 0.0191 |     104 B |
|     ImmutableArray_Builder_ToImmutable |     3 |   102.87 ns |  2.1851 ns |   3.8271 ns | 0.0122 |      64 B |
| ImmutableArray_Builder_MoveToImmutable |     3 |    73.28 ns |  0.7152 ns |   0.5584 ns | 0.0075 |      40 B |
|             ImmutableArray_CreateRange |    30 | 1,860.60 ns | 72.7219 ns | 209.8194 ns | 0.0896 |     472 B |
|     ImmutableArray_Builder_ToImmutable |    30 |   534.11 ns | 11.9213 ns |  24.6195 ns | 0.0525 |     280 B |
| ImmutableArray_Builder_MoveToImmutable |    30 |   492.87 ns | 10.0324 ns |  14.7054 ns | 0.0277 |     148 B |